numalogic/udfs/preprocess.py

```diff
@@ -0,0 +1,194 @@
+        self.pl_conf.stream_confs[config_id] = conf
+
+    def get_conf(self, config_id: str) -> StreamConf:
+        try:
+            return self.pl_conf.stream_confs[config_id]
+        except KeyError as err:
+            raise ConfigNotFoundError(f"Config with ID {config_id} not found!") from err
+
+    def _load_model_from_config(self, preprocess_cfg):
+        preproc_clfs = []
+        for _cfg in preprocess_cfg:
+            _clf = self.preproc_factory.get_instance(_cfg)
+            preproc_clfs.append(_clf)
+        return make_pipeline(*preproc_clfs)
+
+    def exec(self, keys: list[str], datum: Datum) -> Messages:
+        """
+        The preprocess function here receives data from the data source.
+
+        Perform preprocess on the input data.
+
+        Args:
+        -------
+        keys: List of keys
+        datum: Datum object
+
+        Returns
+        -------
+        Messages instance
+
+        """
+        _start_time = time.perf_counter()
+
+        # check message sanity
+        try:
+            data_payload = orjson.loads(datum.value)
+            _LOGGER.info("%s - Data payload: %s", data_payload["uuid"], data_payload)
+        except (orjson.JSONDecodeError, KeyError):  # catch json decode error only
+            _LOGGER.exception("Error while decoding input json")
+            return Messages(Message.to_drop())
+        raw_df, timestamps = get_df(
+            data_payload=data_payload, stream_conf=self.get_conf(data_payload["config_id"])
+        )
+
+        # Drop message if dataframe shape conditions are not met
+        if raw_df.shape[0] < self.get_conf(data_payload["config_id"]).window_size or raw_df.shape[
+            1
+        ] != len(self.get_conf(data_payload["config_id"]).metrics):
+            _LOGGER.error("Dataframe shape: (%f, %f) error ", raw_df.shape[0], raw_df.shape[1])
+            return Messages(Message.to_drop())
+        # Make StreamPayload object
+        payload = make_stream_payload(data_payload, raw_df, timestamps, keys)
+
+        # Check if model will be present in registry
+        if any(
+            [_conf.stateful for _conf in self.get_conf(payload.config_id).numalogic_conf.preprocess]
+        ):
+            preproc_artifact = _load_model(
+                skeys=keys,
+                dkeys=[
+                    _conf.name
+                    for _conf in self.get_conf(payload.config_id).numalogic_conf.preprocess
+                ],
+                payload=payload,
+                model_registry=self.model_registry,
+            )
+            if preproc_artifact:
+                preproc_clf = preproc_artifact.artifact
+                _LOGGER.info(
+                    "%s - Loaded model from: %s",
+                    payload.uuid,
+                    preproc_artifact.extras.get("source"),
+                )
+            else:
+                payload = replace(
+                    payload, status=Status.ARTIFACT_NOT_FOUND, header=Header.TRAIN_REQUEST
```

Comment:
Same again, _load_model function returns None in Exceptions too, skip TRAIN_REQUEST in such cases

Reply:
Added TODO to improve